cmake/extensions.cmake

```diff
@@ -1021,3 +1021,40 @@ function(check_if_directory_is_writeable dir ok)
   endif()
 endfunction()
 
+function(find_appropriate_cache_directory dir)
```

Comment:
It seems better to avoid the environment variable search for the platform-recommended locations unless CMAKE_HOST_SYSTEM_NAME first says we're on that platform. Something like this (relies on HOME always being available on macOS, Linux, and other Unix, which is safe):

```
function(find_appropriate_cache_directory dir)
  # User directory for storing caches, as recommended by the platform.
  set(caches_dir NOTFOUND)

  # Check for standard locations for user-specific cache files on
  # supported platforms.
  if(${CMAKE_HOST_SYSTEM_NAME} STREQUAL Darwin)
    # On macOS, ~/Library/Caches is the preferred cache directory.
    set(caches_dir $ENV{HOME}/Library/Caches)
  elseif(${CMAKE_HOST_SYSTEM_NAME} STREQUAL Windows)
    # On Windows, %LOCALAPPDATA% is recommended.
    if(DEFINED ENV{LOCALAPPDATA})
      set(caches_dir $ENV{LOCALAPPDATA})
    endif()
  elseif(${CMAKE_HOST_SYSTEM_NAME} STREQUAL Linux)
    # On Linux, freedesktop.org recommends using $XDG_CACHE_HOME if
    # that is defined, and defaulting to $HOME/.cache otherwise.
    if(DEFINED ENV{XDG_CACHE_HOME})
      set(caches_dir $ENV{XDG_CACHE_HOME})
    else()
      set(caches_dir $ENV{HOME}/.cache)
    endif()
  elseif(UNIX)
    # On other UNIX-ish operating systems, fall back on ~/.cache.
    set(caches_dir $ENV{HOME}/.cache)
  endif()

  # Fall back to a safe place if the platform directory is not found
  # or not writeable.
  set(fallback_caches_dir ${ZEPHYR_BASE}/.cache)
  if(caches_dir)
    check_if_directory_is_writeable(${caches_dir} ok)
    if(NOT ${ok})
      set(caches_dir ${fallback_caches_dir})
    endif()
  else()
    set(caches_dir ${fallback_caches_dir})
  endif()

  # Zephyr's cache files go in the "zephyr" subdirectory of the user's
  # cache directory.
  set(${dir} ${caches_dir}/zephyr PARENT_SCOPE)
endfunction()
```

Reply:
This seems more robust. @SebastianBoe mind taking a look at the code above?

Reply:
"It seems better to avoid the environment variable search for the platform-recommended locations unless CMAKE_HOST_SYSTEM_NAME first says we're on that platform. "

I have re-written it to comply with this requirement. I made a few different design decisions though. E.g. it appears that for whatever reason CMAKE_HOST_SYSTEM_NAME is not available, but CMAKE_HOST_xxx is. (AFAICT this has something to do with how we cross compile and use project()). Also, I don't trust the platform detection so in case the platform detection fails we fallback to checking HOME and XDG_CACHE_HOME, instead of falling back to not checking any env vars.